### DISCUSSION TITLE
Warning when an cast in foreach loop is necessary

### DISCUSSION THREAD

**Discussion by gatlex** [dotnet/csharplang#7765] [Language Ideas] [answered]
*TL;DR:* I think it would be nice to issue a compiler warning, when the `foreach`-transformation it does involves a _necessary_ explicit cast.

*Long version:* As explained in the [language specification](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#1395-the-foreach-statement), the loop
```c#
foreach (V v in x) «embedded_statement»
```
is transformed into
```c#
{
    E e = ((C)(x)).GetEnumerator();
    try
    {
        while (e.MoveNext())
        {
            V v = (V)(T)e.Current;
            «embedded_statement»
        }
    }
    finally
    {
        ... // Dispose e
    }
}
```
As you can see, there is a deliberate, explicit cast (which is also the reason why [`foreach` can throw an InvalidCastException](https://stackoverflow.com/q/949798/2473926)).

I'm not to much of an expert, but I think the compiler should be able to statically check whether the explicit casts involved are even truly _necessary_. (Maybe this assumption is wrong, in which case I stand corrected.) _If_ so, I suggest to emit a warning, stating that the `foreach` uses an explicit cast under the hood. (This warning can be ignored for compatibility and it might be promoted to an error for the more cautious.)

(This suggestion was spawned from https://stackoverflow.com/a/77653928/2473926.)

**Comment by 333fred** [comment, marked as answer]
Duplicate of https://github.com/dotnet/runtime/issues/48529. There's a lot of good discussions there on why the request was not approved.

**Reply by gatlex** [reply]
Just FTR: This functionality seems to be available as a "code style" check IDE0220: https://stackoverflow.com/a/77654888/2473926. (Even though in my opinion it's more than style but skimming through the discussions this seems quite a controversial topic).
